### PR TITLE
feat(python): Implement `from_dataframe` natively (interchange protocol)

### LINKED ISSUE #10035: Implement Python dataframe interchange protocol natively
### Problem description

The Python dataframe interchange protocol is becoming more and more prevalent. 

For us to be able to interchange with third party libraries, I don't want to have an obligatory dependency on pyarrow.

I think we can revive the old PR made by @stinodego and fill in the missing parts.

As an intermediate solution we can use what worked already from that PR and only require pyarrow for the methods we don't support yet.

### PR DESCRIPTION
**Pull request by stinodego**
Closes #10035 

#### Changes

* Implement `from_dataframe` natively rather than relying on the PyArrow implementation.

This means pyarrow is no longer a required dependency when using `from_dataframe`, and we no longer suffer any limitations of their implementation.

This implementation does everything the previous one did, with some additional benefits:
* (Some) categoricals can be converted zero-copy
* Supports time and duration types
* Better support for various sentinel values for constructing the validity buffer
* Better support for zero-copy in certain edge cases

This implementation has been tested extensively with regular unit tests (100% coverage) as well as a lot of parametric tests for testing integration with pyarrow and pandas. Oddities and bugs are handled and have been reported to the respective projects.

I will probably have missed some edge cases as testing the integration is difficult. However, I am convinced this is worth publishing in its current form!

### CODE REVIEW

File: py-polars/polars/interchange/from_dataframe.py [hunk 1]
```diff
@@ -37,41 +44,277 @@ def from_dataframe(df: SupportsInterchange, *, allow_copy: bool = True) -> DataF
...
+    if not chunks:
+        polars_chunk = _protocol_df_chunk_to_polars(df, allow_copy=allow_copy)
+        chunks.append(polars_chunk)
+
+    return F.concat(chunks, rechunk=False)
+
+
+def _protocol_df_chunk_to_polars(
+    df: InterchangeDataFrame, *, allow_copy: bool = True
+) -> DataFrame:
```
Comment (MarcoGorelli) [inline py-polars/polars/interchange/from_dataframe.py:69]: this is a really small nitpick, but for internal functions, how about not setting defaults? so, just `*, allow_copy: bool)`. Otherwise, it's easy to forget to set it when calling then function (at least, that's something I often do)

File: py-polars/polars/interchange/from_dataframe.py [hunk 2]
```diff
@@ -37,41 +44,277 @@ def from_dataframe(df: SupportsInterchange, *, allow_copy: bool = True) -> DataF
...
+    data = data[offset:]
+
+    # Add the validity buffer if present
+    validity_buffer = _construct_validity_buffer(
+        buffers["validity"], column, String, data, offset, allow_copy=allow_copy
+    )
+    if validity_buffer is not None:
+        data = pl.Series._from_buffers(
+            String, data=data_buffers, validity=validity_buffer
         )
```
Comment (MarcoGorelli) [inline py-polars/polars/interchange/from_dataframe.py:130]: is this right?
```python
df = pd.DataFrame({'a': ['a', 'b', None, 'c', None, 'd']}, dtype='string[pyarrow]')
dfpa = pa.Table.from_pandas(df)
print(pl.from_dataframe(dfpa[1:]))
```

gives

```
Traceback (most recent call last):
  File "/home/marcogorelli/polars-dev/py-polars/t.py", line 12, in <module>
    print(pl.from_dataframe(dfpa[1:]))
  File "/home/marcogorelli/polars-dev/py-polars/polars/convert.py", line 770, in from_dataframe
    return from_dataframe(df, allow_copy=allow_copy)
  File "/home/marcogorelli/polars-dev/py-polars/polars/interchange/from_dataframe.py", line 47, in from_dataframe
    return _from_dataframe(
  File "/home/marcogorelli/polars-dev/py-polars/polars/interchange/from_dataframe.py", line 56, in _from_dataframe
    polars_chunk = _protocol_df_chunk_to_polars(chunk, allow_copy=allow_copy)
  File "/home/marcogorelli/polars-dev/py-polars/polars/interchange/from_dataframe.py", line 74, in _protocol_df_chunk_to_polars
    s = _string_column_to_series(column, allow_copy=allow_copy)
  File "/home/marcogorelli/polars-dev/py-polars/polars/interchange/from_dataframe.py", line 127, in _string_column_to_series
    data = pl.Series._from_buffers(
  File "/home/marcogorelli/polars-dev/py-polars/polars/series/series.py", line 492, in _from_buffers
    return self._from_pyseries(PySeries._from_buffers(dtype, data, validity))
pyo3_runtime.PanicException: called `Result::unwrap()` on an `Err` value: ComputeError(ErrString("validity mask length must match the number of values"))
```

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
awesome that you're working on this and that it's finding so many issues

File: py-polars/polars/interchange/from_dataframe.py (hunk 1 above)
Comment (stinodego) [reply py-polars/polars/interchange/from_dataframe.py:69]: That's fair enough. Will update 👍

File: py-polars/polars/interchange/from_dataframe.py (hunk 2 above)
Comment (stinodego) [reply py-polars/polars/interchange/from_dataframe.py:130]: It's not! I tested chunked strings but not chunked strings with nulls... update incoming.

Comment (MarcoGorelli) [reply py-polars/polars/interchange/from_dataframe.py]: thanks! I remembered a similar one from pandas so it was one of the first things I checked https://github.com/pandas-dev/pandas/issues/52845

I'll take a close look at the rest this week

File: py-polars/polars/interchange/from_dataframe.py [hunk 3]
```diff
@@ -37,41 +44,277 @@ def from_dataframe(df: SupportsInterchange, *, allow_copy: bool = True) -> DataF
...
+            raise CopyNotAllowedError(msg)
+
+        try:
+            sentinel = pl.Series([null_value])
+            if column_dtype.is_temporal():
+                sentinel = sentinel.cast(column_dtype)
+            return data != sentinel  # noqa: TRY300
+        except ComputeError as e:
+            msg = f"invalid sentinel value for column of type {column_dtype}: {null_value!r}"
+            raise TypeError(msg) from e
```
Comment (MarcoGorelli) [inline py-polars/polars/interchange/from_dataframe.py:278]: I think the ruff code here is a good one?
```python
        sentinel = pl.Series([null_value])
        if column_dtype.is_temporal():
            try:
                sentinel = sentinel.cast(column_dtype)
            except ComputeError as e:
                msg = f"invalid sentinel value for column of type {column_dtype}: {null_value!r}"
                raise TypeError(msg) from e
        return data != sentinel
```
so that the `try` block is as small and atomic as possible

File: py-polars/tests/unit/interchange/test_roundtrip.py [hunk 4]
```diff
@@ -109,3 +241,13 @@ def test_roundtrip_pyarrow_boolean_midbyte_slice() -> None:
...
+
+
+@pytest.mark.xfail(
+    reason="Bug in pandas: https://github.com/pandas-dev/pandas/issues/56712"
+)
+def test_from_dataframe_pandas_timestamp_ns() -> None:
+    df = pl.Series("a", [datetime(2000, 1, 1)], dtype=pl.Datetime("ns")).to_frame()
+    df_pd = df.to_pandas(use_pyarrow_extension_array=True)
+    result = pl.from_dataframe(df_pd)
+    assert_frame_equal(result, df)
```
Comment (MarcoGorelli) [inline py-polars/tests/unit/interchange/test_roundtrip.py:253]: this is fixed on `main`, and 2.2 should be out soon (this month) - simplest thing might be to just wait until it's out? else `xfail` based on the pandas version?

File: py-polars/polars/interchange/from_dataframe.py (hunk 3 above)
Comment (stinodego) [reply py-polars/polars/interchange/from_dataframe.py:278]: Both the `cast` and the equality operation can cause a ComputeError if the sentinel value is invalid. So I have them both in the try block. Placing the sentinel in a Series shouldn't fail, so I'll extract that one.

File: py-polars/tests/unit/interchange/test_roundtrip.py (hunk 4 above)
Comment (stinodego) [reply py-polars/tests/unit/interchange/test_roundtrip.py]: I made it xfail based on the pandas version 👍

File: py-polars/tests/unit/interchange/test_roundtrip.py [hunk 5]
```diff
@@ -44,42 +49,170 @@ def test_roundtrip_pyarrow_parametric(df: pl.DataFrame) -> None:
...
+        excluded_dtypes=[
+            pl.Categorical,  # Categoricals come back as Enums
+            # large string not yet supported by pandas
+            # https://github.com/pandas-dev/pandas/issues/56702
+            pl.String,
+            pl.Float32,  # NaN values come back as nulls
+            pl.Float64,  # NaN values come back as nulls
+            # pandas exports nanosecond pyarrow types incorrectly
+            # https://github.com/pandas-dev/pandas/issues/56712
+            pl.Datetime("ns"),
```
Comment (MarcoGorelli) [inline py-polars/tests/unit/interchange/test_roundtrip.py:130]: note to self, this can be removed after the next pandas release

File: py-polars/tests/unit/interchange/test_roundtrip.py [hunk 6]
```diff
@@ -44,42 +49,170 @@ def test_roundtrip_pyarrow_parametric(df: pl.DataFrame) -> None:
...
+            pl.Categorical,  # Categoricals come back as Enums
+            # large string not yet supported by pandas
+            # https://github.com/pandas-dev/pandas/issues/56702
+            pl.String,
+            pl.Float32,  # NaN values come back as nulls
+            pl.Float64,  # NaN values come back as nulls
+            pl.Boolean,  # pandas exports boolean buffers as byte-packed
+            # pandas exports nanosecond pyarrow types incorrectly
+            # https://github.com/pandas-dev/pandas/issues/56712
+            pl.Datetime("ns"),
```
Comment (MarcoGorelli) [inline py-polars/tests/unit/interchange/test_roundtrip.py:157]: same

**Review by MarcoGorelli — APPROVED** [review]
Awesome stuff!

Got a little question (and a todo note about the timestamp('ns') bug, but that can happen later), but agree that this would be good to get in

File: py-polars/tests/unit/interchange/test_from_dataframe.py [hunk 7]
```diff
@@ -56,77 +107,461 @@ def test_from_dataframe_pyarrow_table_zero_copy() -> None:
...
+
+    col = PatchableColumn(s)
+    col.describe_null = (ColumnNullType.USE_BITMASK, 0)
+    col.null_count = 2
+
+    dtype = (DtypeKind.BOOL, 1, "b", NE)
+    validity_buffer_info = (bitmask, dtype)
+
+    result = _construct_validity_buffer(
+        validity_buffer_info, col, s.dtype, s, allow_copy=True
```
[the remark this answers is not in the document]
Comment (stinodego) [reply py-polars/tests/unit/interchange/test_from_dataframe.py:434]: I could set it to `False` but it doesn't really matter for this code path.

File: py-polars/tests/unit/interchange/test_roundtrip.py (hunk 5 above)
Comment (stinodego) [reply py-polars/tests/unit/interchange/test_roundtrip.py:130]: I'll go back and update the tests after pandas releases (remove the xfail, update some of the parametrics).

**Comment by stinodego** [comment]
Thanks for the review Marco! Much appreciated. I'll go ahead with this and let's see if we get some feedback from users.
